src/vos/pmdk_log.c

```diff
@@ -0,0 +1,100 @@
+/**
+ * (C) Copyright 2024 Intel Corporation.
+ * (C) Copyright 2025 Hewlett Packard Enterprise Development LP
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * Connecting the PMDK's logging to DAOS logging.
+ *
+ * vos/pmdk_log.c
+ */
+#define D_LOGFAC DD_FAC(pmdk)
+
+#ifdef DAOS_PMEM_BUILD
+#include <daos/debug.h>
+#include <daos/common.h>
+#include <libpmemobj/log.h>
+#include <libpmemobj.h>
+
+static uint64_t pmemobj_log_level_2_dlog_prio[] = {
+    [PMEMOBJ_LOG_LEVEL_HARK] = DLOG_INFO,   [PMEMOBJ_LOG_LEVEL_FATAL] = DLOG_CRIT,
+    [PMEMOBJ_LOG_LEVEL_ERROR] = DLOG_ERR,   [PMEMOBJ_LOG_LEVEL_WARNING] = DLOG_WARN,
+    [PMEMOBJ_LOG_LEVEL_NOTICE] = DLOG_NOTE, [PMEMOBJ_LOG_LEVEL_INFO] = DLOG_INFO,
+    [PMEMOBJ_LOG_LEVEL_DEBUG] = DLOG_DBG,
+};
+
+static void
+pmdk_log_function(enum pmemobj_log_level level, const char *file_name, unsigned line_no,
+		  const char *function_name, const char *message)
+{
+	uint64_t dlog_prio = pmemobj_log_level_2_dlog_prio[level];
```

Comment:
This is one example where pulling in PMDK into DAOS would be useful.

Anyway, this can be improved to be more efficient.

Add the following

`#define _D_DEBUG_EXPANDED(func, flag, flag_var, ...)`

Have _D_DEBUG call this with DD_FLAG(flag, D_LOGFAC) as the flag var.

In your case, you can have a map like the one you have above to each flag var

e.g. &DD_FLAG(DLOG_INFO, D_LOGFAC)

And instead of calling _D_DEBUG, call _D_DEBUG_EXPANDED with *flag_var you saved.

Additionally, one might consider passing these variables down to the lower level (e..g. register_debug_callback(callback, &DD_FLAG(...)).  The check can just be if (*flag != 0) inside of pmdk and the upper level will take care of caching the value for the next call.

Reply:
The `_D_DEBUG_W_SAVED_MASK` macro (instead of ` _D_DEBUG_EXPANDED`) has been added and the `pmdk_log_function()` has been adjusted.